### PR TITLE
Filterable and groupable CDS

### LINKED ISSUE #4070: Plotting with subsets (row-wise) of ColumnDataSources (like filtered data)
Currently, Bokeh only allows glyphs to take full columns of data from a ColumnDataSource. This means that you must have additional ColumnDataSources for any plots that only use a subset of the data. This causes data to be replicated and makes linked selection more complicated for users, who would have to keep track of indices themselves. 

It would be nice if Bokeh allowed glyphs to take subsets of columns as data. An application using multiple plots, each showing some subset (e.g. filtered by the value of some column, like the example @dennisobrien showed at the SF meetup) of the data (or the full set), would then be able to completely share the data in the same way that Bokeh allows plots to share the full ColumnDataSource now. Linked selection would work intuitively with selected data points (rows) showing up in whichever plots contain that point. Holding shift before selecting on any plot would add to the selection. 

One possibility for an API would be to include another parameter that picks the subset of data for a renderer to use by the indices (which could default to all). 

``` python
data_source = ColumnDataSource(df)
fig = figure()
r = fig.circle(x = ‘x’, y = ‘y’, source=data_source, indices=[0, 1, 2]) 
r = fig.circle(x = ‘x’, y = ‘y’, source=data_source, indices=np.arange(n)[df.weather == “sunny”])
r = fig.circle(x = ‘x’, y = ‘y’, source=data_source, indices=[i for i, val in enumerate(df.weather) if val == “sunny”])
```

Another possibility could be to make the data source “filterable” in a pandas-like way. 

``` python
r = fig.circle(x = ‘x’, y = ‘y’, source=data_source[ data_source.weather == “sunny” ]) 
```

I tried to look through the issues and google group to find some related posts. 
https://github.com/bokeh/bokeh/issues/2323
https://github.com/bokeh/bokeh/issues/2710
https://groups.google.com/a/continuum.io/forum/#!searchin/bokeh/data$20source%7Csort:date/bokeh/_DIOQGlC6ds/EbZL0KRfBwAJ
https://groups.google.com/a/continuum.io/forum/#!searchin/bokeh/data$20source%7Csort:date/bokeh/79eSMaDV39I/OTfZqZHIug0J

Is this something that would be in Bokeh’s scope?

### PR DESCRIPTION
**Pull request by clairetang6**
This PR is a second attempt to create a filterable CDS. The first attempt in [this PR](https://github.com/bokeh/bokeh/pull/5828) created a new data source called the TableDataSource. This second attempt instead puts the filtering functionality into the ColumnDataSource (CDS). 

In this approach, the data and filters are separated by pulling out the CDS's data into a new model called the ColumnDataStore (data store). The data store would be responsible for holding the data as well as the selection on the full data set. (It would also be responsible for streaming and patching the data and the CDS could have stream and patch methods that called the data store's stream and patch methods.)

I additionally added a group property to the CDS, which specifies a column name and a group name to restrict to. This allows for an API that looks like (described by @bryevdv [here](https://github.com/bokeh/bokeh/issues/5884#issuecomment-280669400))

```python
fig.circle('sepal_width', 'petal_width', source=source, group=('species', 'setosa'))
```

With the first two commits in this PR, you can have something like:

iris_groups

```python
from bokeh.plotting import figure
from bokeh.layouts import row, column
from bokeh.models.sources import ColumnDataSource
from bokeh.models.data_stores import ColumnDataStore
from bokeh.io import output_file, show

from bokeh.sampledata.iris import flowers

output_file("filter_group_cds.html")

import numpy as np

data_store = ColumnDataStore(flowers)
source0 = ColumnDataSource(data_store)
source1 = ColumnDataSource(data_store, filter=lambda cds: np.array(cds.data_store.data['species']) == 'setosa')
source2 = ColumnDataSource(data_store, group=['species', 'versicolor'])
source3 = ColumnDataSource(data_store, group=['species', 'virginica'])

plot_size_and_tools = {'plot_height': 300, 'plot_width': 300,
                        'tools':['box_select', 'reset']}

fig1 = figure(title="Full data set", **plot_size_and_tools)
fig1.circle(x='petal_length', y='petal_width', source=source0, color='black')

fig2 = figure(title="Composed of filtered and grouped data", x_range=fig1.x_range, y_range=fig1.y_range, **plot_size_and_tools)
fig2.circle(x='petal_length', y='petal_width', source=source1, color='red')
fig2.square(x='petal_length', y='petal_width', source=source2, color='green')
fig2.diamond(x='petal_length', y='petal_width', source=source3, color='blue')

show(row(fig1, fig2))
```

One issue shown the GIF above is that selecting points on the right figure only works with the last renderer. This issue is discussed in [issue 2136](https://github.com/bokeh/bokeh/issues/2136) and [issue 2989](https://github.com/bokeh/bokeh/issues/2989).

This PR is still in early stages, and I'm looking for feedback at any level. The group property may not be strictly necessary. Users can filter as in source1 in the above example for a specific group on their own and a higher-level plot.scatter method (with a groupby parameter) could use filter internally as well. 

With this PR, each filter is represented through one CDS. Another approach could be to have a CDS contain a mapping of renderers to filters. Then, each renderer could use the same CDS. The filter could be made on the creation of a renderer and be added to the CDS's renderer-filter mapping. 

- [ ] issues: fixes #4070
- [ ] tests added / passed
- [ ] release document entry (if new feature or API change)

### CODE REVIEW

**Comment by bryevdv** [comment]
My initial concern is that we will have to move the data store functionality up into CDS somehow. We have been demonstrating `source.data = new_data` for years now, in both JS and python, and breaking all that code would be too disruptive.

**Comment by bryevdv** [comment]
I have some further thoughts on this. Perhaps we can discuss in a call. A very high level idea description would be this:

Leave CDS basically alone. Make filters/groups be things *external* to a CDS that can provide a *view* on the data that the CDS contains. Then things like glyph renderers can be updated to operate either directly over a CDS, or if a "view" is configured, to operate over the view that wraps the CDS. Ideally we can define something akin to an "iteration protocol" that both CDS and views both have, then things that need to iterate over data don't have to care which one they are iterating over. 

In other words, instead of pushing the current CDS "down" into another object `ColumnDataStore` that is contained by a new and too-different CDS, leave the current CDS as-is and elevate views to be something above it (i.e. that can be configured with a CDS)

**Comment by clairetang6** [comment]
That sounds similar to the first approach I took where I created a new model (TableDataSource) that contained a reference to a CDS. 

What I am trying now as a third approach is to have a mapping of filters-to-renderers stored by a CDS. Then the data for the full data set is still contained in the CDS and .data will be unchanged. If a glyph renderer has an associated filter in that mapping, then they can use that filter.

**Comment by bryevdv** [comment]
> That sounds similar to the first approach I took where I created a new model (TableDataSource) that contained a reference to a CDS.

Yes the thought occurred to me, it might be a matter of naming (names are important!) I think it's important to establish these things as *optional views* on a *data source*, rather than a new kind of at a source. And I think a unified iteration protocol is also important (I don't recall if the other PR did that or not, offhand)

**Comment by clairetang6** [comment]
Superseded by [5982](https://github.com/bokeh/bokeh/pull/5982)
